// lesslint/lib/rule/require-after-space.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

'use strict';

/**
 * 规则名称
 *
 * @const
 * @type {string}
 */
/**
 * @file
 * `:` : 1. 属性名后的冒号（:）与属性值之间必须（MUST）保留一个空格，冒号前不得（MUST NOT）保留空格。
 *       2. 定义变量时冒号（:）与变量值之间必须（MUST）保留一个空格，冒号前不得（MUST NOT）保留空格。
 * `,` : 1. 在用逗号（,）分隔的列表（Less 函数参数列表、以 , 分隔的属性值等）中，逗号后必须（MUST）保留一个空格，
 *      逗号前不得（MUST NOT）保留空格。
 *       2. 在给 mixin 传递参数时，在参数分隔符（, / ;）后必须（MUST）保留一个空格
 *
 * 逗号暂时不太好实现
 *
 * https://github.com/ecomfe/spec/blob/master/less-code-style.md#%E5%B1%9E%E6%80%A7%E5%8F%98%E9%87%8F
 * @author ielgnaw(wuji0223@gmail.com)
 */

var RULENAME = 'require-after-space';

/**
 * 冒号
 *
 * @const
 * @type {string}
 */
var COLON = ':';

/**
 * 逗号
 *
 * @const
 * @type {string}
 */
var COMMA = ',';

/**
 * 匹配 css 属性值的 url(...);
 *
 * @const
 * @type {RegExp}
 */
var PATTERN_URI = /url\(["']?([^\)"']+)["']?\)/i;

/**
 * 冒号的错误信息
 *
 * @const
 * @type {string}
 */
var COLON_MSG = '' + 'Disallow contain spaces between the `attr-name` and `:`, ' + 'Must contain spaces between `:` and `attr-value`';

/**
 * 逗号的错误信息
 *
 * @const
 * @type {string}
 */
var COMMA_MSG = 'Must contain spaces after `,` in `attr-value`';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        var ruleVal = opts.ruleVal;
        var realRuleVal = [];
        Array.prototype.push[Array.isArray(ruleVal) ? 'apply' : 'call'](realRuleVal, ruleVal);

        if (realRuleVal.length) {

            css.walkDecls(function (decl) {
                var source = decl.source;
                var line = source.start.line;
                var lineContent = (0, _util.getLineContent)(line, source.input.css) || '';

                if (realRuleVal.indexOf(COLON) !== -1) {
                    var between = decl.raws.between;

                    if (between.slice(0, 1) !== ':' // `属性名` 与之后的 `:` 之间包含空格了
                    || between.slice(-1) === ':' // `:` 与 `属性值` 之间不包含空格
                    ) {
                            var colorStr = decl.prop + decl.raws.between + decl.value;
                            result.warn(RULENAME, {
                                node: decl,
                                ruleName: RULENAME,
                                line: line,
                                message: COLON_MSG,
                                colorMessage: '`' + lineContent.replace(colorStr, _chalk2.default.magenta(colorStr)) + '` ' + _chalk2.default.grey(COLON_MSG)
                            });
                        }
                }

                if (realRuleVal.indexOf(COMMA) !== -1) {
                    var value = decl.value;

                    // 排除掉 uri 的情况，例如
                    // background-image: url(data:image/gif;base64,R0lGODlhCAAHAIABAGZmZv...);
                    // background-image: 2px 2px url(data:image/gif;base64,R0lGODlhCAAHAIABAGZmZv...);
                    // background-image: url(data:image/gif;base64,R0lGODlhCAAHAIABAGZmZv...) 2px 2px;
                    if (!PATTERN_URI.test(value)) {
                        var items = lineContent.split(';');
                        for (var j = 0, jLen = items.length; j < jLen; j++) {
                            var s = items[j];
                            if (s.indexOf(',') > -1 && /.*,(?!\s)/.test(s) && s.length !== lineContent.length // s.length === lineContent.length 的情况表示当前行结束了
                            ) {
                                    result.warn(RULENAME, {
                                        node: decl,
                                        ruleName: RULENAME,
                                        errorChar: COMMA,
                                        line: line,
                                        message: COMMA_MSG,
                                        colorMessage: '`' + lineContent.replace(value, _chalk2.default.magenta(value)) + '` ' + _chalk2.default.grey(COMMA_MSG)
                                    });
                                }
                        }
                    }
                }
            });
        }
    };
});
