// lesslint/lib/cli.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.parse = parse;

var _fs = require('fs');

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _edpCore = require('edp-core');

var _package = require('../package');

var _package2 = _interopRequireDefault(_package);

var _util = require('./util');

var _checker = require('./checker');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file 命令行功能模块
 * @author ielgnaw(wuji0223@gmail.com)
 */

'use strict';

/**
 * 显示默认的信息
 */
var showDefaultInfo = function showDefaultInfo() {
    console.warn(_package2.default);
    console.log('');
    console.log(_package2.default.name + ' v' + _package2.default.version);
    console.log(_chalk2.default.bold.green((0, _util.formatMsg)(_package2.default.description)));
};

/**
 * 校验结果报告
 *
 * @inner
 * @param {Object} errors 按文件类型为 key，值为对应的校验错误信息列表的对象
 */
var report = function report(errors) {
    var t12 = true;

    if (errors.length) {
        errors.forEach(function (error) {
            _edpCore.log.info(error.path);
            error.messages.sort(function (left, right) {
                return left.line - right.line;
            });
            error.messages.forEach(function (message) {
                var ruleName = message.ruleName || '';
                var msg = '→ ' + (ruleName ? _chalk2.default.bold(ruleName) + ': ' : '');
                // 全局性的错误可能没有位置信息
                if (typeof message.line === 'number') {
                    msg += 'line ' + message.line;
                    if (typeof message.col === 'number') {
                        msg += ', col ' + message.col;
                    }
                    msg += ': ';
                }

                msg += message.colorMessage || message.message;
                _edpCore.log.warn(msg);
            });
        });
        t12 = false;
    }

    if (t12) {
        _edpCore.log.info('Congratulations! Everything gone well, you are T12!');
    } else {
        process.exit(1);
    }
};

/**
 * 解析参数。作为命令行执行的入口
 *
 * @param {Array} args 参数列表
 */
function parse(args) {
    args = args.slice(2);

    // 不带参数时，默认检测当前目录下所有的 less 文件
    if (args.length === 0) {
        args.push('.');
    }

    if (args[0] === '--version' || args[0] === '-v') {
        showDefaultInfo();
        return;
    }

    var patterns = ['**/*.less', '!**/{output,test,node_modules,asset,dist,release,doc,dep,report}/**'];

    var candidates = (0, _util.getCandidates)(args, patterns);

    var count = candidates.length;

    if (!count) {
        return;
    }

    // 错误信息的集合
    var errors = [];

    /**
     * 每个文件的校验结果回调，主要用于统计校验完成情况
     *
     * @inner
     */
    var callback = function callback() {
        count--;
        if (!count) {
            report(errors);
        }
    };

    // 遍历每个需要检测的 less 文件
    candidates.forEach(function (candidate) {
        var readable = (0, _fs.createReadStream)(candidate, {
            encoding: 'utf8'
        });
        readable.on('data', function (chunk) {
            var file = {
                content: chunk,
                path: candidate
            };
            (0, _checker.check)(file, errors, callback);
        });
        readable.on('error', function (err) {
            throw err;
        });
    });
}
