// csshint/lib/rule/bulletproof-font-face.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'bulletproof-font-face'; /**
                                         * @file bulletproof-font-face 的检测逻辑
                                         *       Rule: Use the bulletproof @font-face syntax to avoid 404's in old IE
                                         *       (http://www.fontspring.com/blog/the-new-bulletproof-font-face-syntax)
                                         *       https://github.com/CSSLint/csslint/wiki/Bulletproof-font-face
                                         * @author ielgnaw(wuji0223@gmail.com)
                                         */

var PATTERN = /^\s?url\(['"].+\.eot\?.*['"]\)\s*format\(['"]embedded-opentype['"]\)[\s\S]*$/i;

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
var MSG = '@font-face declaration doesn\'t follow the fontspring bulletproof syntax';

var firstSrc = true;
var failedDecl = false;

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {

        if (opts.ruleVal) {
            css.walkAtRules(function (atRule) {
                if (atRule.name !== 'font-face') {
                    return;
                }

                firstSrc = true;
                failedDecl = false;

                atRule.walkDecls(function (decl) {
                    if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                        return;
                    }

                    var prop = decl.prop,
                        value = decl.value;


                    if (prop === 'src') {
                        if (!value.match(PATTERN) && firstSrc) {
                            failedDecl = decl;
                            firstSrc = false;
                        } else if (value.match(PATTERN) && !firstSrc) {
                            failedDecl = false;
                        }
                    }
                });

                if (failedDecl) {
                    var source = failedDecl.source;
                    var line = source.start.line;
                    var lineContent = (0, _util.getLineContent)(line, source.input.css);
                    var col = source.start.column;
                    result.warn(RULENAME, {
                        node: atRule,
                        ruleName: RULENAME,
                        line: line,
                        col: col,
                        message: '`' + lineContent + '` ' + MSG,
                        colorMessage: '`' + (0, _util.changeColorByStartAndEndIndex)(lineContent, col, source.end.column) + '` ' + _chalk2.default.grey(MSG)
                    });

                    global.CSSHINT_INVALID_ALL_COUNT++;
                }
            });
        }
    };
});
