// csshint/lib/rule/disallow-overqualified-elements.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'disallow-overqualified-elements';

/**
 * css 组合的正则匹配
 *
 * @const
 * @type {RegExp}
 */
/**
 * @file disallow-overqualified-elements 的检测逻辑
 *       013: [强制] 如无必要，不得为 `id`、`class` 选择器添加类型选择器进行限定。
 * @author ielgnaw(wuji0223@gmail.com)
 */

var PATTERN_COMBINATORS = /[\s>+~,[]+/;

/**
 * css selector 开始字符的正则匹配
 *
 * @const
 * @type {RegExp}
 */
var PATTERN_STARTCHARS = /[\.#\[]+/;

var PATTERN_PERCENT = /^((-|\+)?\d{1,2}(\.\d+)?|100)%$/;

/**
 * 错误信息
 *
 * @const
 * @type {string}
 */
var MSG = 'Not allowed to add a type selector is limited to ID, class selector';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (opts.ruleVal) {

            css.walkRules(function (rule) {

                if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                    return;
                }

                if (!isNaN(rule.selector) || PATTERN_PERCENT.test(rule.selector)) {
                    return;
                }

                var source = rule.source;
                var line = source.start.line;

                var lineContent = (0, _util.getLineContent)(line, source.input.css) || '';

                var segments = rule.selector.split(PATTERN_COMBINATORS);
                for (var i = 0, len = segments.length; i < len; i++) {
                    var items = segments[i].split(PATTERN_STARTCHARS);
                    if (items[0] !== '' && items.length > 1) {
                        result.warn(RULENAME, {
                            node: rule,
                            ruleName: RULENAME,
                            line: line,
                            message: MSG,
                            colorMessage: '`' + lineContent.replace(segments[i], segments[i].replace(items[0], _chalk2.default.magenta(items[0]))) + '` ' + _chalk2.default.grey(MSG)
                        });
                        global.CSSHINT_INVALID_ALL_COUNT++;
                    }
                }
            });
        }
    };
});
