// csshint/lib/rule/duplicate-background-images.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'duplicate-background-images';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file duplicate-background-images 的检测逻辑
 *       Every background-image should be unique. Use a common class for e.g. sprites
 *       https://github.com/CSSLint/csslint/wiki/Disallow-duplicate-background-images
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'Every background-image should be unique. Use a common class for e.g. sprites';

/**
 * 匹配 css 中 url 的正则
 *
 * @const
 * @type {RegExp}
 */
var PATTERN_URL = /\burl\s*\((["']?)([^\)]+)\1\)/g;

var stack = {};

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {

        if (!opts.ruleVal) {
            return;
        }

        stack = {};

        css.walkDecls(function (decl) {

            var prop = decl.prop;
            if (prop.match(/background/i)) {
                var value = decl.value;
                var match = null;
                /* eslint-disable no-extra-boolean-cast */
                while (!!(match = PATTERN_URL.exec(value))) {
                    if (typeof stack[match[2]] === 'undefined') {
                        stack[match[2]] = decl;
                    } else {
                        var str = 'Background image `' + match[2] + '` was used multiple times, first declared at line ' + stack[match[2]].source.start.line + ', col ' + stack[match[2]].source.start.column + '. ' + MSG;

                        var colorStr = 'Background image `' + _chalk2.default.magenta(match[2]) + '` was used multiple times, first declared at line ' + stack[match[2]].source.start.line + ', col ' + stack[match[2]].source.start.column + '. ' + _chalk2.default.grey(MSG);

                        var source = decl.source;
                        var line = source.start.line;
                        var lineContent = (0, _util.getLineContent)(line, source.input.css);
                        var col = lineContent.indexOf(match[2]) + 1;
                        result.warn(RULENAME, {
                            node: decl,
                            ruleName: RULENAME,
                            line: line,
                            col: col,
                            message: str,
                            colorMessage: colorStr
                        });
                        global.CSSHINT_INVALID_ALL_COUNT++;
                    }
                }
                /* eslint-enable no-extra-boolean-cast */
            }
        });
    };
});
