// csshint/lib/rule/ids.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'ids';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file ids 的检测逻辑
 *       Selectors should not contain IDs
 *       https://github.com/CSSLint/csslint/wiki/Disallow-IDs-in-selectors
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'Selectors should not contain IDs';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {

        if (!opts.ruleVal) {
            return;
        }

        css.walkRules(function (rule) {
            if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                return;
            }

            var selector = rule.selector,
                source = rule.source;

            var selectorGroup = selector.split(',');
            var line = source.start.line;
            var col = source.start.column;
            var lineContent = (0, _util.getLineContent)(line, source.input.css);

            for (var i = 0, len = selectorGroup.length; i < len; i++) {
                var selectorInGroup = selectorGroup[i] || '';
                // 去掉 attr 选择器
                selectorInGroup = selectorInGroup.replace(/\[.+?\](?::[^\s>+~\.#\[]+)?/g, '');
                var match = selectorInGroup.match(/#[^\s>+~\.#\[]+/);
                if (match) {
                    if (selectorInGroup.slice(0, 1) === '\n') {
                        line = line + 1;
                        lineContent = (0, _util.getLineContent)(line, source.input.css);
                        col = col + match.index - 1;
                    } else {
                        col = col + match.index;
                    }
                    result.warn(RULENAME, {
                        node: rule,
                        ruleName: RULENAME,
                        line: line,
                        col: col,
                        message: MSG,
                        colorMessage: '`' + lineContent.replace(match[0], _chalk2.default.magenta(match[0])) + '` ' + _chalk2.default.grey(MSG)
                    });
                }
            }
        });
    };
});
