// csshint/lib/rule/leading-zero.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'leading-zero';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file leading-zero 的检测逻辑
 *       025: [强制] 当数值为 0 - 1 之间的小数时，省略整数部分的 `0`。
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'When value is between 0 - 1 decimal, omitting the integer part of the `0`';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (opts.ruleVal) {

            if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                return;
            }

            css.walkDecls(function (decl) {
                var parts = _postcss2.default.list.space(decl.value);
                var source = decl.source;
                var line = source.start.line;
                for (var i = 0, len = parts.length; i < len; i++) {
                    var part = parts[i];
                    var numericVal = parseFloat(part);
                    if (numericVal < 1 && numericVal > 0) {
                        if (part.slice(0, 2) === '0.') {
                            var lineContent = (0, _util.getLineContent)(line, source.input.css);
                            result.warn(RULENAME, {
                                node: decl,
                                ruleName: RULENAME,
                                line: line,
                                col: lineContent.indexOf(part) + 1,
                                message: MSG,
                                colorMessage: '`' + (0, _util.changeColorByIndex)(lineContent, lineContent.indexOf(part), part) + '` ' + _chalk2.default.grey(MSG)
                            });
                            global.CSSHINT_INVALID_ALL_COUNT++;
                        }
                    }
                }
            });
        }
    };
});
