// csshint/lib/rule/omit-protocol-in-url.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'omit-protocol-in-url';

/**
 * 匹配 css 中 url 的正则
 *
 * @const
 * @type {RegExp}
 */
/**
 * @file omit-protocol-in-url 的检测逻辑
 *       027: [建议] `url()` 函数中的绝对路径可省去协议名。
 * @author ielgnaw(wuji0223@gmail.com)
 */

var PATTERN_URL = /\burl\s*\((["']?)([^\)]+)\1\)/g;

/**
 * 匹配 url() 中 path 的协议
 *
 * @const
 * @type {RegExp}
 */
var PATTERN_PROTOCOL = /^((https?|s?ftp|irc[6s]?|git|afp|telnet|smb):\/\/){1}/gi;

/**
 * 错误信息
 *
 * @const
 * @type {string}
 */
var MSG = 'Path in the `url()` should remove protocol';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (opts.ruleVal) {

            css.walkDecls(function (decl) {

                if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                    return;
                }

                var source = decl.source,
                    value = decl.value;

                var line = source.start.line;
                var lineContent = (0, _util.getLineContent)(line, source.input.css);

                var match = null;
                var matchProtocol = null;

                /* eslint-disable no-extra-boolean-cast */
                while (!!(match = PATTERN_URL.exec(value))) {
                    var url = match[2];

                    // decl.value 相对于 lineContent 的 index
                    var valueIndex = lineContent.indexOf(decl.value);

                    // 相对于 decl.value 的 index
                    var index = valueIndex + match.input.indexOf(url);
                    while (!!(matchProtocol = PATTERN_PROTOCOL.exec(url))) {
                        result.warn(RULENAME, {
                            node: decl,
                            ruleName: RULENAME,
                            line: line,
                            col: index + 1,
                            message: MSG,
                            colorMessage: '`' + (0, _util.changeColorByIndex)(lineContent, index, matchProtocol[0]) + '` ' + _chalk2.default.grey(MSG)
                        });
                        global.CSSHINT_INVALID_ALL_COUNT++;
                    }
                }
                /* eslint-enable no-extra-boolean-cast */
            });
        }
    };
});
