// csshint/lib/rule/property-not-existed.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _prefixes = require('../prefixes');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var prefixList = (0, _prefixes.getPrefixList)();

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
/**
 * @file property-not-existed 的检测逻辑，检测属性是否存在
 * @author ielgnaw(wuji0223@gmail.com)
 */

var RULENAME = 'property-not-existed';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {

        if (opts.ruleVal) {

            css.walkDecls(function (decl) {
                if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                    return;
                }

                var prop = decl.prop;
                var standardProperty = prop.replace(/^\-(webkit|moz|ms|o)\-/g, '');
                // 标准模式在 prefixList 中，那么如果 propertyName 不在 prefixList 中
                // 即这个属性用错了，例如 -o-animation
                if (prefixList.indexOf(standardProperty) > -1) {
                    if (prefixList.indexOf(prop) <= -1) {

                        var source = decl.source;
                        var line = source.start.line;
                        var col = source.start.column;

                        result.warn(RULENAME, {
                            node: decl,
                            ruleName: RULENAME,
                            line: line,
                            col: col,
                            message: '' + 'Current property ' + '`' + prop + '` ' + 'is not existed',
                            colorMessage: '' + 'Current property ' + '`' + _chalk2.default.magenta(prop) + '` ' + 'is not existed'
                        });
                        global.CSSHINT_INVALID_ALL_COUNT++;
                    }
                }
            });
        }
    };
});
