// csshint/lib/rule/text-indent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'text-indent';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file text-indent 的检测逻辑
 *       Checks for text indent less than -99px
 *       https://github.com/CSSLint/csslint/wiki/Disallow-negative-text-indent
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = '' + 'Negative text-indent doesn\'t work well with RTL.' + 'If you use text-indent for image replacement explicitly set direction for that item to ltr';

var textIndentDecl = void 0;
var direction = void 0;

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (!opts.ruleVal) {
            return;
        }

        css.walkRules(function (rule) {

            textIndentDecl = false;
            direction = 'inherit';

            rule.walkDecls(function (decl) {
                if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                    return;
                }
                var prop = decl.prop;
                var value = (0, _util.getPropertyValue)(decl.value);

                if (prop === 'text-indent' && value[0].value < -99) {
                    textIndentDecl = decl;
                } else if (prop === 'direction' && value.value === 'ltr') {
                    direction = 'ltr';
                }
            });

            if (textIndentDecl && direction !== 'ltr') {
                var source = textIndentDecl.source;
                var line = source.start.line;
                var lineContent = (0, _util.getLineContent)(line, source.input.css);
                var col = source.start.column;
                result.warn(RULENAME, {
                    node: rule,
                    ruleName: RULENAME,
                    line: line,
                    col: col,
                    message: MSG,
                    colorMessage: '`' + lineContent.replace(textIndentDecl.prop, _chalk2.default.magenta(textIndentDecl.prop)) + '` ' + _chalk2.default.grey(MSG)
                });
                global.CSSHINT_INVALID_ALL_COUNT++;
            }
        });
    };
});
