// csshint/lib/rule/unique-headings.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'unique-headings';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file unique-headings 的检测逻辑
 *       Headings should be defined only once
 *       https://github.com/CSSLint/csslint/wiki/Headings-should-only-be-defined-once
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'Headings should be defined only once';

/**
 * css 组合的正则匹配
 *
 * @const
 * @type {RegExp}
 */
var PATTERN_COMBINATORS = /[\s>+~]+/g;

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {

        if (!opts.ruleVal) {
            return;
        }

        var headings = {
            h1: 0,
            h2: 0,
            h3: 0,
            h4: 0,
            h5: 0,
            h6: 0
        };

        css.walkRules(function (rule) {
            /* jshint maxstatements: 26 */
            if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                return;
            }

            var selector = rule.selector,
                source = rule.source;

            var selectorGroup = selector.split(',');
            var line = source.start.line;
            var col = source.start.column;
            var lineContent = (0, _util.getLineContent)(line, source.input.css);

            for (var i = 0, len = selectorGroup.length; i < len; i++) {
                var selectorInGroup = selectorGroup[i] || '';
                var segments = selectorInGroup.split(PATTERN_COMBINATORS);
                var segmentLen = segments.length;

                var lastSegment = segments[segmentLen - 1];
                if (!lastSegment.match(':') && headings.hasOwnProperty(lastSegment)) {
                    headings[lastSegment]++;
                    if (headings[lastSegment] > 1) {
                        var newLineMatch = selectorInGroup.match(/\n/g);
                        var extraLine = 0;
                        if (newLineMatch) {
                            extraLine += newLineMatch.length;
                            line = line + extraLine;
                            lineContent = (0, _util.getLineContent)(line, source.input.css);
                            col = col + lineContent.indexOf(lastSegment);
                        } else {
                            col = lineContent.indexOf(lastSegment) + 1;
                        }
                        result.warn(RULENAME, {
                            node: rule,
                            ruleName: RULENAME,
                            line: line,
                            col: col,
                            message: MSG,
                            colorMessage: '`' + lineContent.replace(lastSegment, _chalk2.default.magenta(lastSegment)) + '` ' + _chalk2.default.grey(MSG)
                        });

                        global.CSSHINT_INVALID_ALL_COUNT++;
                    }
                }
            }
        });
    };
});
